common/state_machine.hpp

```diff
@@ -0,0 +1,231 @@
+
+private:
+	struct TransitionCondition {
+		// Note: Comparing address-of states. We don't want to rely on comparison operators
+		// in the states themselves, and we just want to know if they are different
+		// instances.
+		State<ContextType, EventType> *state;
+		EventType event;
+
+		bool operator==(const TransitionCondition &t) const {
+			return state == t.state && event == t.event;
+		}
+	};
+
+	class Hasher {
+	public:
+		size_t operator()(const TransitionCondition &obj) const {
+			return std::hash<State<ContextType, EventType> *>()(obj.state)
+				   ^ std::hash<int>()(static_cast<int>(obj.event));
+		}
+	};
+
+	State<ContextType, EventType> *current_state_;
+	unordered_map<TransitionCondition, State<ContextType, EventType> *, Hasher> transitions_;
+	unordered_set<EventType> deferred_events_;
+
+	friend class StateMachineRunner<ContextType, EventType>;
+
+public:
+	void AddTransition(
+		State<ContextType, EventType> &source_state,
+		EventType event,
+		State<ContextType, EventType> &target_state,
+		TransitionFlag flag) {
+		transitions_[TransitionCondition {&source_state, event}] = &target_state;
+		if (flag == TransitionFlag::Deferred) {
+			// Event is involved in at least one deferred transition, so mark that.
+			deferred_events_.insert(event);
+		}
+	}
+};
+
+template <typename ContextType, typename EventType>
+class StateMachineRunner : virtual public EventPoster<EventType> {
+public:
+	StateMachineRunner(ContextType &ctx) :
+		ctx_(ctx) {
+	}
+	StateMachineRunner(StateMachineRunner &) = delete;
+
+	void PostEvent(EventType event) override {
+		event_queue_.push(event);
+		PostToEventLoop();
+	}
+
+	void RunOne() {
+		if (event_queue_.empty()) {
+			return;
+		}
+
+		size_t size = event_queue_.size();
+		vector<State<ContextType, EventType> *> to_run;
+
+		for (size_t count = 0; count < size; count++) {
+			bool deferred = false;
+			auto event = event_queue_.front();
+			event_queue_.pop();
+			to_run.clear();
+
+			for (auto machine : machines_) {
+				typename StateMachine<ContextType, EventType>::TransitionCondition cond {
+					machine->current_state_, event};
+				if (machine->deferred_events_.find(event) != machine->deferred_events_.end()) {
+					deferred = true;
+				}
+
+				auto match = machine->transitions_.find(cond);
+				if (match == machine->transitions_.end()) {
+					// No match in this machine, continue.
+					continue;
+				}
+
+				auto &target = match->second;
+				to_run.push_back(target);
+				machine->current_state_ = target;
+			}
+
+			if (to_run.empty()) {
+				if (deferred) {
+					// Put back in the queue to try later. This won't be tried
+					// again during this run, due to only making `size`
+					// attempts in the for loop.
+					event_queue_.push(event);
+				} else {
+					log::Warning(
+						"State machine event " + to_string(static_cast<int>(event))
+						+ " was not handled by any state. This is a bug and could hang the state machine.");
+					assert(!to_run.empty());
```

Comment:
Might as well do `assert(false)` here because it's inside an `if` with an exactly opposite condition.

Reply:
I know, I left it in mostly as "documentation". It's debug mode only anyway, so I think it's fine to leave it.